### PR TITLE
add new api's

### PR DESCRIPTION
**Pull request by remstone7**
This PR does the following:
Version 1.0.6

- [x] get's  the current extension version `reclaim()`

- [x] returns product by item_id and quote_id, returns images

- [x] returns products by id range

- [x] Adds klaviyo.js

- [x] Add server side track ability

TODO 

- [x] Fix product id return on product-inspector

- [ ] Add auth to api endpoints

product endpoint
filter 1 is quote_id and filter_2 is item_id
image

Version endpoint
image

product-inspector endpoint:
filter_1 start id filter_2 end id
image

### CODE REVIEW

File: view/frontend/templates/analytics/initialize.phtml [hunk 1]
```diff
@@ -1,24 +1,14 @@
 <?php if ( $this->isKlaviyoEnabled() && $this->getPublicApiKey() ): ?>
+<script type="application/javascript" async 
+ src="https://static.klaviyo.com/onsite/js/klaviyo.js?company_id=<?php echo $this->getPublicApiKey(); ?>"></script>
```
Comment (dmitrymrn) [inline view/frontend/templates/analytics/initialize.phtml]: It looks like `getPublicApiKey()` (and potentially other methods) perform DB requests. It means every page load introduces an extra load. As far as I remember, Magento has cache and it should be possible to cache these values.

File: Model/Reclaim.php [hunk 2]
```diff
@@ -38,4 +69,61 @@ public function stores()
...
     }
+    public function product($filter) {
+        $quote_id = $filter['1'];
+        $item_id = $filter['2'];
+        $quote = $this->quoteFactory->create()->load($quote_id);
+        
+        if (!$quote){
+            return array('error' => 'not found');
+        }
+        $item = $quote->getItemById($item_id);
```
Comment (klaviyojad) [inline Model/Reclaim.php]: maybe throw an error as well if that item_id wasnt invalid

File: Model/Reclaim.php [hunk 3]
```diff
@@ -38,4 +69,61 @@ public function stores()
...
+        return $response;
+
+    }
+    public function _getImages($product){
+        $images = $product->getMediaGalleryImages();
+        $image_array = array();
+        foreach($images as $image) {
+            $image_url = $image->getUrl();
+            if ($image_url){
+                array_push($image_array, $image_url);
```
Comment (dmitrymrn) [inline Model/Reclaim.php]: or you could do (totally optional)
```suggestion
                $image_array[] = $image_url;
```

File: Model/Reclaim.php [hunk 4]
```diff
@@ -38,4 +69,61 @@ public function stores()
...
+            if (!$product){
+                continue;
+            }
+            $data = array(
+                'id' => $product->getProductId(),
+                'sku' => $product->getSku(),
+                'type_id' => $product->getTypeId(),
+                'price' => $product->getPrice(),
+            );
+            array_push($response, $data);
```
Comment (dmitrymrn) [inline Model/Reclaim.php]: Unlike Python, PHP passes by value and not by reference. This means that every new variable takes extra memory. While it's good to be explicit in the Python world, in PHP it comes with a cost. Here is a more efficient pattern:
```suggestion
            $response[] = array(
                'id' => $product->getProductId(),
                'sku' => $product->getSku(),
                'type_id' => $product->getTypeId(),
                'price' => $product->getPrice()
            )
```

File: Model/Reclaim.php [hunk 5]
```diff
@@ -38,4 +69,61 @@ public function stores()
 
         return $hydrated_stores;
     }
+    public function product($filter) {
+        $quote_id = $filter['1'];
+        $item_id = $filter['2'];
```
Comment (dmitrymrn) [inline Model/Reclaim.php]: Dumb question, but why Magento makes indexes to be strings?..

File: Model/Reclaim.php [hunk 6]
```diff
@@ -38,4 +69,61 @@ public function stores()
...
+        
+        if (!$quote){
+            return array('error' => 'not found');
+        }
+        $item = $quote->getItemById($item_id);
+        $product = $this->_objectManager->get('Magento\Catalog\Model\Product')->load($item->getProductId());
+
+        $image_array = $this->_getImages($product);
+        
+        $response['$body'] = [
```
Comment (dmitrymrn) [inline Model/Reclaim.php]: is `'$body'` a special Klaviyo var?

Comment (remstone7) [reply Model/Reclaim.php]: removed, no reason for it

File: Model/Reclaim.php (hunk 5 above)
Comment (remstone7) [reply Model/Reclaim.php]: it's a filter from the url..

http://magento2.local:8888/rest/V1/klaviyo/reclaim/product?filter[1]=4&filter[2]=15

it takes the queryparams and filter['1'] is grabbing the key of the filter

Comment (dmitrymrn) [reply Model/Reclaim.php]: image

Comment (Jaeto) [reply Model/Reclaim.php]: Do we have control over the query param keys? Having to tie to `filter[1]` and `filter[2]` is rough.

Comment (remstone7) [reply Model/Reclaim.php]: nice find, updated

**Pull request merged**